leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java

```diff
@@ -225,18 +224,19 @@ protected Object handleInvocation(Object proxy, Method method, Object[] args) th
             return null;
         }
 
-        final LeadershipToken leadershipToken = getLeadershipToken();
+        ListenableFuture<LeadershipToken> leadershipToken = getLeadershipToken();
```

Comment:
I don't know you'd want this - it seems reasonable to throw if you're knowably not the leader

Reply:
but does add some complexity?

Reply:
Idk - it's not something i'd like to be part of the contract, if that makes sense

Reply:
Yep, the main reason for this was that I didn't want to throw in the async case. I did consider the alternative of splitting here to decide whether to throw immediately or return immediateFailedFuture on whether we were on an async path - didn't really like either solution. No strong opinions, wouldn't mind changing if you think that works better.

It does seem that we should handle the exceptions here similar to how we get them in the `isStillLeading` part (that could just be that I know my round is not the latest, for instance), which I would imagine is throwing for sync and returning a failed future for async.